### PR TITLE
Help mode

### PR DESCRIPTION
**Pull request by yangruihan**
Hello, this is an amazing project, I fall in love with it.

I've added a "help" mode to show the AI calculating strategy results instead of executing them directly, to help players who just need hints, I've also added support for mahjong emoji to make it show better, new AI strategy branches I've also added support, but you can wait until you're done developing it before merging it in.

截屏2022-05-03 下午2 07 52

### CODE REVIEW

**Comment by yangruihan** [comment]
And I also add some yaku calculate logic in `yaku.js`

File: src/main.js [hunk 1]
```diff
@@ -92,18 +96,60 @@ function main() {
...
 		setTimeout(main, 1000);
+
+		if (MODE === AIMODE.HELP) {
+			oldOps = [];
+		}
+
 		return;
 	}
 
-	currentActionOutput.value = "Calculating best move...";
```
Comment (Jimboom7) [inline src/main.js:100]: I think moving this from here into the mainOwnTurn function is not going to work. The browser won't refresh the GUI element while the javascript thread is blocking it. Only if this is called before the setTimeout() the browser has time to update it.

File: src/yaku.js [hunk 2]
```diff
@@ -63,7 +68,8 @@ function getYaku(inputHand, inputCalls) {
 	//Sanankou
 	//3 concealed triplets
 	//Open*
-	var sanankou = getSanankou(inputHand);
+	let sanankou = getSanankou(triplets);
```
Comment (Jimboom7) [inline src/yaku.js:71]: This is not going to work: `triplets` cointains also the triplets from calls/melds. That means by using this variable the function will also check for open triplets which is basically the same as Toitoi.

File: src/yaku.js [hunk 3]
```diff
@@ -155,37 +171,63 @@ function getYaku(inputHand, inputCalls) {
...
-	var daisangen = getDaisangen(hand);
+	let daisangen = getDaisangen(hand);
+	//log(`-- daisangen open ${daisangen.open} ${daisangen.closed}`)
 	yakuOpen += daisangen.open;
 	yakuClosed += daisangen.closed;
 
 	//Suuankou
 	//4 Concealed Triplets
 	//Closed
+	let suuankou = getSuuankou(triplets);
```
Comment (Jimboom7) [inline src/yaku.js:182]: Same problem as with Sanankou here.

File: src/yaku.js [hunk 4]
```diff
@@ -317,14 +358,97 @@ function getShousangen(hand) {
...
+//Suushiihou
+function getSuushiihou(hand) {
+	if (hand.filter(tile => tile.type === 3 && tile.index <= 4).length == 11) {
+		return { open: YAKUMAN_SCORE, closed: YAKUMAN_SCORE };
+	}
+	return { open: 0, closed: 0 };
+}
+
+//ChuurenPoutou
+function getChuurenPoutou(hand, triplets, sequences) {
```
Comment (Jimboom7) [inline src/yaku.js:421]: In general it would be nice to have some testcases for these yaku, but I can add these myself later.

File: src/utils.js [hunk 5]
```diff
@@ -655,3 +655,69 @@ function binomialCoefficient(a, b) {
...
+		case 9: return "ron";
+		case 10: return "kyuushu kyuuhai";
+		case 11: return "kita";
+		default: return type;
+	}
+}
+
+//Get Emoji str by tile name
+function getTileEmoji(name) {
+	switch (name) {
```
Comment (Jimboom7) [inline src/utils.js:678]: I think putting the Emojis into a 2-dimensional array (10x4) would make this function way shorter. Something like this:
```
var emojiList = [
["red🀋", "🀇", "🀈", ........],
[........],
[........],
[........]];
return emojiList[tile.type][tile.index];
```

**Review by Jimboom7 — CHANGES_REQUESTED** [review]
I like this very much :)
I only looked at the code for now and it looks good. But before I merge it I need to test it in actual games to make sure everything is still working.

**Comment by yangruihan** [comment]
Of course, this is a must, and I am currently testing.

File: src/yaku.js (hunk 2 above)
Comment (yangruihan) [reply src/yaku.js:71]: Yes, but this PR is only about the "help" mode, so I will temporarily remove this part of the logic, I will test this part of the logic more thoroughly before making another PR. ;-)

File: src/utils.js (hunk 5 above)
Comment (yangruihan) [reply src/utils.js:678]: That sounds good, I will change it!

File: src/yaku.js (hunk 4 above)
Comment (yangruihan) [reply src/yaku.js:421]: Remove it.

File: src/yaku.js (hunk 3 above)
Comment (yangruihan) [reply src/yaku.js:182]: Remove it.

File: src/main.js (hunk 1 above)
Comment (yangruihan) [reply src/main.js:100]: I revert this code back and it looks better now.

**Comment by yangruihan** [comment]
Add `.gitignore` file to ignore build directory.

**Comment by dioarame** [comment]
@yangruihan Thanks for Help GUI design.
But I have an opinion of this GUI.
As my experience of testing GUI, not just tile code but Image of tile is pretty good source.
But, Source image is so little to indicate in game situation and console log.
So, How about modify like this? => [Tile Image : 7z]
Then in many of situation, it'll be more look good I think.

**Comment by yangruihan** [comment]
> @yangruihan Thanks for Help GUI design.
> 
> But I have an opinion of this GUI.
> 
> As my experience of testing GUI, not just tile code but Image of tile is pretty good source.
> 
> But, Source image is so little to indicate in game situation and console log.
> 
> So, How about modify like this? => [Tile Image : 7z]
> 
> Then in many of situation, it'll be more look good I think.

You are right, I have thought about this too, since we have the same idea, I will add this part sometime.

**Comment by yangruihan** [comment]
> @yangruihan Thanks for Help GUI design. But I have an opinion of this GUI. As my experience of testing GUI, not just tile code but Image of tile is pretty good source. But, Source image is so little to indicate in game situation and console log. So, How about modify like this? => [Tile Image : 7z] Then in many of situation, it'll be more look good I think.

I have made a new PR https://github.com/Jimboom7/AlphaJong/pull/50, you can go and see if it meets your expectations, thanks for the feedback.

**Pull request merged**